ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto

```diff
@@ -0,0 +1,53 @@
+//
+// Protocol buffer messages used by the participant state key-value utilities
+// for implementing a DAML ledger backed by a key-value store.
+//
+
+syntax = "proto3";
+package com.daml.ledger.participant.state.kvutils;
+option java_package = "com.daml.ledger.participant.state.kvutils";
+
+import "google/protobuf/timestamp.proto";
+import "daml-lf/transaction/src/main/protobuf/transaction.proto";
+
+// The ledger implementation stores the `DamlKVEntry`s in an
+// ordered sequence of keys, allowing for offset-based querying.
+//
+// The value in which entry (or entries) is stored should include
+// the record time at which it was committed.
+//
+// Each entry can be converted into a participant state `Update` event.
+message DamlKVEntry {
```

Comment:
Reading the later code I'm seeing an alternative structure for how we partition the key-value space and store the state of the DAML Ledger.

The key idea is to represent the state of the ledger as a `([DamlLogEntry],Map DamlStateKey DamlStateValue)` tuple. A transactions inputs can both refer to past log entries via their `Int` offset and to `DamlStateKey` bytes. Transaction validation is given the current recordTime and the values associated to the requested inputs. It must return the modification to the ledger as a `([DamlLogEntry], Map DamlStateKey DamlStateValue)` tuple as well. The semantics is to append the `DamlLogEntry`s and to overwrite the given keys with the new values.

Note that in this case a failed validation should result in a command-rejection log entry; i.e., the ledger is always extended with the transaction, but it's state modification depends on its validity.

The benefit of this scheme is that we can evolve from different tx-representations and kv-representations in a backwards compatible form by extending the protobufs. In particular, we can smoothly switch to a decomposed representation of transactions.

Reply:
Also: `validation` should be able to fail fatally with `Left "reason"` to denote a missing input or the like.

Reply:
sounds reasonable! I'll do a quick draft and then verify with you.